File: C#/WindowsPhone/LGame-XNA-lib/Loon.Action.Avg/AVGAnm.cs
namespace Loon.Action.Avg
{
    using System;
    using System.IO;
    using System.Collections.Generic;
    using Loon.Core.Geom;
    using Loon.Core;
    using Loon.Core.Graphics.Opengl;
    using Loon.Core.Graphics;
    using Loon.Core.Resource;
    using Loon.Utils;
    using Loon.Java;
    using Loon.Action.Avg.Drama;

    public class AVGAnm : Expression, LRelease
    {

        internal readonly Point.Point2i point = new Point.Point2i();

        private string path;

        internal float alpha = 1.0f;

        internal float angle;

        internal int width, height, imageWidth, imageHeight;

        internal List<Int32> posxTmps = new List<Int32>();

        internal List<Int32> posyTmps = new List<Int32>();

        internal int[] posx = null;

        internal int[] posy = null;

        internal List<Int32> time = new List<Int32>();

        internal int tmp_time = 20;

        internal int alltime = 0;

        internal int count = 0;

        internal long startTime = -1;

        internal bool loop = true, load = false;

        internal LTexture texture;

        internal LColor color;

        public AVGAnm(string resName)
            : this(Resources.OpenStream(resName))
        {
        }

        public AVGAnm(Stream ins0)
        {
            Open(ins0);
        }

        public void Open(Stream ins0)
        {
            try
            {
                StreamReader reader = new StreamReader(ins0, System.Text.Encoding.UTF8);
                string script = null;
                for (; (script = reader.ReadLine()) != null; )
                {
                    if (script.Length > 0 && !script.StartsWith(FLAG_L_TAG)
                            && !script.StartsWith(FLAG_C_TAG)
                            && !script.StartsWith(FLAG_I_TAG))
                    {
                        string[] element = StringUtils.Split(script, ";");
                        for (int j = 0; j < element.Length; j++)
                        {
                            Load(element[j]);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                this.load = false;
                Loon.Utils.Debugging.Log.Exception(ex);
                return;
            }
            this.load = true;
            this.count = posxTmps.Count;
            this.posx = new int[count];
            this.posy = new int[count];
            for (int i = 0; i < count; i++)
            {
                this.posx[i] = (int)(posxTmps[i]);
                this.posy[i] = (int)(posyTmps[i]);
            }
            if (width == 0)
            {
                width = imageWidth;
            }
            if (height == 0)
            {
                height = imageHeight;
            }
        }

        private void Load(string script)
        {
            string[] op = StringUtils.Split(script, "=");
            if (op.Length == 2)
            {
                string key = op[0].Trim();
                string value_ren = op[1].Trim();
                if ("path".Equals(key, StringComparison.InvariantCultureIgnoreCase))
                {
                    path = value_ren.Replace("\"", "");
                    if (texture != null)
                    {
                        texture.Destroy();
                        texture = null;
                    }
                    if (GLEx.Self != null)
                    {
                        texture = new LTexture(path);
                        imageWidth = texture.GetWidth();
                        imageHeight = texture.GetHeight();
                    }
                }
                if ("imagewidth".Equals(key, StringComparison.InvariantCultureIgnoreCase))
                {
                    if (MathUtils.IsNan(value_ren))
                    {
                        imageWidth = Int32.Parse(value_ren);
                    }
                }
                else if ("alpha".Equals(key, StringComparison.InvariantCultureIgnoreCase))
                {
                    if (MathUtils.IsNan(value_ren))
                    {
                        alpha = Single.Parse(value_ren, JavaRuntime.NumberFormat);
                    }
                }
                else if ("angle".Equals(key, StringComparison.InvariantCultureIgnoreCase))
                {
                    if (MathUtils.IsNan(value_ren))
                    {
                        angle = Single.Parse(value_ren, JavaRuntime.NumberFormat);
                    }
                }
                else if ("color".Equals(key, StringComparison.InvariantCultureIgnoreCase))
                {
                    string[] p = StringUtils.Split(value_ren, ",");
                    if (p.Length > 2 && p.Length < 5)
                    {
                        for (int i = 0; i < p.Length; i++)
                        {
                            p[i] = p[i].Replace("^[\\t ]*", "").Replace(
                                    "[\\t ]*$", "");
                        }
                        if (p.Length == 3)
                        {
                            color = new LColor(Int32.Parse(p[0]),
                                    Int32.Parse(p[1]), Int32.Parse(p[2]));
                        }
                        if (p.Length == 4)
                        {
                            color = new LColor(Int32.Parse(p[0]),
                                    Int32.Parse(p[1]), Int32.Parse(p[2]),
                                    Int32.Parse(p[3]));
                        }
                    }
                }
                else if ("imageheight".Equals(key, StringComparison.InvariantCultureIgnoreCase))
                {
                    if (MathUtils.IsNan(value_ren))
                        imageHeight = Int32.Parse(value_ren);
                }
                else if ("width".Equals(key, StringComparison.InvariantCultureIgnoreCase))
                {
                    if (MathUtils.IsNan(value_ren))
                    {
                        width = Int32.Parse(value_ren);
                    }
                }
                else if ("height".Equals(key, StringComparison.InvariantCultureIgnoreCase))
                {
                    if (MathUtils.IsNan(value_ren))
                    {
                        height = Int32.Parse(value_ren);
                    }
                }
                else if ("time".Equals(key, StringComparison.InvariantCultureIgnoreCase))
                {
                    if (MathUtils.IsNan(value_ren))
                    {
                        tmp_time = Int32.Parse(value_ren);
                    }
                }
                else if ("pos".Equals(key, StringComparison.InvariantCultureIgnoreCase))
                {
                    string[] p_0 = StringUtils.Split(value_ren, ",");
                    for (int i_1 = 0; i_1 < p_0.Length; i_1++)
                    {
                        p_0[i_1] = p_0[i_1].Replace("^[\\t ]*", "").Replace(
                                "[\\t ]*$", "");
                    }
                    switch (p_0.Length)
                    {
                        case 1:
                            if (MathUtils.IsNan(p_0[0]))
                            {
                                CollectionUtils.Add(posxTmps, Int32.Parse(p_0[0]));
                                CollectionUtils.Add(posyTmps, Int32.Parse(p_0[0]));
                                CollectionUtils.Add(time, tmp_time);
                                alltime += tmp_time;
                            }
                            break;
                        case 2:
                            if (MathUtils.IsNan(p_0[0]) && MathUtils.IsNan(p_0[1]))
                            {
                                CollectionUtils.Add(posxTmps, Int32.Parse(p_0[0]));
                                CollectionUtils.Add(posyTmps, Int32.Parse(p_0[1]));
                                CollectionUtils.Add(time, tmp_time);
                                alltime += tmp_time;
                            }
                            break;
                    }
                }
            }
        }

        public string GetPath()
        {
            return path;
        }

        public int GetWidth()
        {
            return width;
        }

        public void SetWidth(int width_0)
        {
            this.width = width_0;
        }

        public int GetHeight()
        {
            return height;
        }

        public void SetHeight(int height_0)
        {
            this.height = height_0;
        }

        public int GetImageWidth()
        {
            return imageWidth;
        }

        public void SetImageWidth(int imageWidth_0)
        {
            this.imageWidth = imageWidth_0;
        }

        public int GetImageHeight()
        {
            return imageHeight;
        }

        public void SetImageHeight(int imageHeight_0)
        {
            this.imageHeight = imageHeight_0;
        }

        public int[] GetPosx()
        {
            return posx;
        }

        public int[] GetPosy()
        {
            return posy;
        }

        public List<Int32> GetTime()
        {
            return time;
        }

        public int GetAlltime()
        {
            return alltime;
        }

        public int GetCount()
        {
            return count;
        }

        public bool IsLoop()
        {
            return loop;
        }

        public LTexture GetTexture()
        {
            return texture;
        }

        public float GetAlpha()
        {
            return alpha;
        }

        public float GetAngle()
        {
            return angle;
        }

        public void Start(long elapsedTime, bool loop_0)
        {
            this.startTime = elapsedTime;
            this.loop = loop_0;
            if (texture != null)
            {
                if (texture.GetWidth() == imageWidth
                        && texture.GetHeight() == imageHeight)
                {
                    this.loop = false;
                }
            }
        }

        public void Start()
        {
            this.Start(0, loop);
        }

        public void Stop()
        {
            this.startTime = -1;
            this.loop = false;
        }

        public Point.Point2i GetPos(long elapsedTime)
        {
            if (startTime != -1)
            {
                int frame = GetFrame(elapsedTime);
                point.x = posx[frame];
                point.y = posy[frame];
            }
            return point;
        }

        private int GetFrame(long elapsedTime)
        {
            long diff = elapsedTime - startTime;
            if (!loop && diff >= alltime)
            {
                startTime = -1;
                return 0;
            }
            long now = diff % alltime;
            int t = 0;
            for (int i = 0; i < count; i++)
            {
                t += time[i];
                if (now < t)
                {
                    return i;
                }
            }
            return 0;
        }

        public void Dispose()
        {
            Stop();
            if (texture != null)
            {
                texture.Destroy();
                texture = null;
            }
        }

    }

}

File: Java/Android/OpenGLES/source/core/src/loon/action/avg/AVGAnm.java
package loon.action.avg;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;

import loon.action.avg.drama.Expression;
import loon.core.LRelease;
import loon.core.LSystem;
import loon.core.geom.Point.Point2i;
import loon.core.graphics.LColor;
import loon.core.graphics.opengl.GLEx;
import loon.core.graphics.opengl.LTexture;
import loon.core.resource.Resources;
import loon.utils.MathUtils;


/**
 * Copyright 2008 - 2010
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.3.3
 */
//0.3.3新增类,用以按指定的格式循环播放动画图像
public class AVGAnm implements Expression, LRelease {

	final Point2i point = new Point2i();

	private String path;

	float alpha = 1.0f;

	float angle;

	int width, height, imageWidth, imageHeight;

	ArrayList<Integer> posxTmps = new ArrayList<Integer>();

	ArrayList<Integer> posyTmps = new ArrayList<Integer>();

	int[] posx = null;

	int[] posy = null;

	ArrayList<Integer> time = new ArrayList<Integer>();

	int tmp_time = 20;

	int alltime = 0;

	int count = 0;

	long startTime = -1;

	boolean loop = true, load = false;

	LTexture texture;

	LColor color;

	public AVGAnm(String resName) throws IOException {
		this(Resources.openResource(resName));
	}

	public AVGAnm(InputStream in) {
		open(in);
	}

	public void open(InputStream in) {
		try {
			BufferedReader reader = new BufferedReader(new InputStreamReader(
					in, LSystem.encoding));
			String script = null;
			for (; (script = reader.readLine()) != null;) {
				if (script.length() > 0 && !script.startsWith(FLAG_L_TAG)
						&& !script.startsWith(FLAG_C_TAG)
						&& !script.startsWith(FLAG_I_TAG)) {
					final String[] element = script.split(";");
					for (int j = 0; j < element.length; j++) {
						load(element[j]);
					}
				}
			}
		} catch (Exception ex) {
			this.load = false;
			ex.printStackTrace();
			return;
		}
		this.load = true;
		this.count = posxTmps.size();
		this.posx = new int[count];
		this.posy = new int[count];
		for (int i = 0; i < count; i++) {
			this.posx[i] = (posxTmps.get(i));
			this.posy[i] = (posyTmps.get(i));
		}
		if (width == 0) {
			width = imageWidth;
		}
		if (height == 0) {
			height = imageHeight;
		}
	}

	private void load(String script) {
		String[] op = script.split("=");
		if (op.length == 2) {
			final String key = op[0].trim();
			final String value = op[1].trim();
			if ("path".equalsIgnoreCase(key)) {
				path = value.replaceAll("\"", "");
				if (texture != null) {
					texture.destroy();
					texture = null;
				}
				if (GLEx.self != null) {
					texture = new LTexture(path);
					imageWidth = texture.getWidth();
					imageHeight = texture.getHeight();
				}
			}
			if ("imagewidth".equalsIgnoreCase(key)) {
				if (MathUtils.isNan(value)) {
					imageWidth = Integer.parseInt(value);
				}
			} else if ("alpha".equalsIgnoreCase(key)) {
				if (MathUtils.isNan(value)) {
					alpha = Float.parseFloat(value);
				}
			} else if ("angle".equalsIgnoreCase(key)) {
				if (MathUtils.isNan(value)) {
					angle = Float.parseFloat(value);
				}
			} else if ("color".equalsIgnoreCase(key)) {
				String[] p = value.split(",");
				if (p.length > 2 && p.length < 5) {
					for (int i = 0; i < p.length; i++) {
						p[i] = p[i].replaceAll("^[\\t ]*", "").replaceAll(
								"[\\t ]*$", "");
					}
					if (p.length == 3) {
						color = new LColor(Integer.parseInt(p[0]),
								Integer.parseInt(p[1]), Integer.parseInt(p[2]));
					}
					if (p.length == 4) {
						color = new LColor(Integer.parseInt(p[0]),
								Integer.parseInt(p[1]), Integer.parseInt(p[2]),
								Integer.parseInt(p[3]));
					}
				}
			} else if ("imageheight".equalsIgnoreCase(key)) {
				if (MathUtils.isNan(value))
					imageHeight = Integer.parseInt(value);
			} else if ("width".equalsIgnoreCase(key)) {
				if (MathUtils.isNan(value)) {
					width = Integer.parseInt(value);
				}
			} else if ("height".equalsIgnoreCase(key)) {
				if (MathUtils.isNan(value)) {
					height = Integer.parseInt(value);
				}
			} else if ("time".equalsIgnoreCase(key)) {
				if (MathUtils.isNan(value)) {
					tmp_time = Integer.parseInt(value);
				}
			} else if ("pos".equalsIgnoreCase(key)) {
				String[] p = value.split(",");
				for (int i = 0; i < p.length; i++) {
					p[i] = p[i].replaceAll("^[\\t ]*", "").replaceAll(
							"[\\t ]*$", "");
				}
				switch (p.length) {
				case 1:
					if (MathUtils.isNan(p[0])) {
						posxTmps.add(Integer.parseInt(p[0]));
						posyTmps.add(Integer.parseInt(p[0]));
						time.add(tmp_time);
						alltime += tmp_time;
					}
					break;
				case 2:
					if (MathUtils.isNan(p[0]) && MathUtils.isNan(p[1])) {
						posxTmps.add(Integer.parseInt(p[0]));
						posyTmps.add(Integer.parseInt(p[1]));
						time.add(tmp_time);
						alltime += tmp_time;
					}
					break;
				}
			}
		}
	}

	public String getPath() {
		return path;
	}

	public int getWidth() {
		return width;
	}

	public void setWidth(int width) {
		this.width = width;
	}

	public int getHeight() {
		return height;
	}

	public void setHeight(int height) {
		this.height = height;
	}

	public int getImageWidth() {
		return imageWidth;
	}

	public void setImageWidth(int imageWidth) {
		this.imageWidth = imageWidth;
	}

	public int getImageHeight() {
		return imageHeight;
	}

	public void setImageHeight(int imageHeight) {
		this.imageHeight = imageHeight;
	}

	public int[] getPosx() {
		return posx;
	}

	public int[] getPosy() {
		return posy;
	}

	public ArrayList<Integer> getTime() {
		return time;
	}

	public int getAlltime() {
		return alltime;
	}

	public int getCount() {
		return count;
	}

	public boolean isLoop() {
		return loop;
	}

	public final LTexture getTexture() {
		return texture;
	}

	public float getAlpha() {
		return alpha;
	}

	public float getAngle() {
		return angle;
	}

	public void start(long elapsedTime, boolean loop) {
		this.startTime = elapsedTime;
		this.loop = loop;
		if (texture != null) {
			if (texture.getWidth() == imageWidth
					&& texture.getHeight() == imageHeight) {
				this.loop = false;
			}
		}
	}

	public void start() {
		this.start(0, loop);
	}

	public void stop() {
		this.startTime = -1;
		this.loop = false;
	}

	public Point2i getPos(long elapsedTime) {
		if (startTime != -1) {
			int frame = getFrame(elapsedTime);
			point.x = posx[frame];
			point.y = posy[frame];
		}
		return point;
	}

	private int getFrame(long elapsedTime) {
		long diff = elapsedTime - startTime;
		if (!loop && diff >= alltime) {
			startTime = -1;
			return 0;
		}
		long now = diff % alltime;
		int t = 0;
		for (int i = 0; i < count; i++) {
			t += time.get(i);
			if (now < t) {
				return i;
			}
		}
		return 0;
	}

	@Override
	public void dispose() {
		stop();
		if (texture != null) {
			texture.destroy();
			texture = null;
		}
	}

}
